contents/scala/Scala.md

```diff
@@ -0,0 +1,284 @@
+```
+object MyExtensions {
+  implicit class IntParity(i: Int) {
+    def isEven = i % 2 == 0
+    def isOdd  = !isEven
+  }
+}
+
+import MyExtensions._  // bring implicit enrichment into scope
+4.isEven  // -> true
+4.isOdd   // -> false
+```
+This is the implicit class that extend the library of class Int. The name of the class does not matter. 
+After importing this class, we can use `isEven()` and `isOdd()` method just as if it has been declared in its own class. 
+
+
+### Pattern Matching
+Scala has built-in support for **pattern matching**, which can be thought of as a more extensible version of a **switch** 
+statement, where arbitrary data types can be matched. The `match` operator is used to do **pattern matching** .
+
+There are many kinds of pattern matching in Scala, let's start with a simplest pattern matching function:
+```
+def matchTest(x: Int): String = x match {
+  case 0 => "zero"
+  case 1 => "one"
+  case 2 => "two"
+  case _ => "many"
+}
+
+matchTest(0)  // zero
+matchTest(3)  // many
+matchTest(1)  // one
+```
+The parameter x is the left operand of the `match` operator and on the right is an expression with four cases. Each case 
+expression is tried to see if it will match, and the first match will be returned. The last case _ is a "catch all" case 
+which can match all inputs.
+
+The **pattern matching** in Scala also includes Case Class Matching, List Matching, Type Matching, Tuple Matching and etc. 
+You can see [PATTERN MATCHING](https://docs.scala-lang.org/tour/pattern-matching.html) for more details.
+
+Following is an example of QuickSort algorithm to show how strong is Scala's pattern matching:
+
+```
+def qsort(list: List[Int]): List[Int] = list match {
+  case Nil => Nil
+  case pivot :: tail =>
+    val (smaller, rest) = tail.partition(_ < pivot)
+    qsort(smaller) ::: pivot :: qsort(rest)
+}
+```
+
+The idea here is that we *partition* a list recursively, sort each part and combine the results together.(See [QuickSort](https://en.wikipedia.org/wiki/Quicksort) for the details of the algorithm)
+
+In this case, `Nil` only matches the object `Nil` (Empty list).
+But `pivot :: tail` matches a non-empty list, and it will destructure the list according to the pattern given. 
+In this case, the code will have a variable `pivot` holding the head of the list, and another variable `tail`
+ holding the tail of the list.
+ 
+In local variable declarations, tuple matching is used. The return value of the call to `tail.partition` is a 
+tuple. In this case, the code will have a variable `smaller` holding the first element of the tuple, and another 
+variable `rest` holding the second element of the list. 
+
+## Getting Started
+There are mainly two ways to work in Scala.
+
+1. Using the **INTELLIJ IDE**. Refer to the [instructions](https://docs.scala-lang.org/getting-started-intellij-track/getting-started-with-scala-in-intellij.html) for more details.
```

Comment:
Intellij is the only IDE for scala? doesn't sound right. In any case, give a learning path in this section i.e., in your opinion, what exact path should the reader follow to learn Scala?